src/common/utils/formatter/formatter.ts

```diff
@@ -139,22 +140,78 @@ export function getFormattedClause(dimension: Dimension, clause: FilterClause, t
       if (clause.action === Filter.CONTAINS) values = `"${values}"`;
 
       break;
-
     case 'time':
-      var selection = (clause.selection as LiteralExpression);
-      var selectionType = selection.type;
-      if (selectionType === 'TIME_RANGE') {
-        var timeRange = selection.value as TimeRange;
-        values = formatTimeRange(timeRange, timezone, DisplayYear.IF_DIFF);
-      } else if (selectionType === "SET/TIME") {
-        values = clauseSet.toString();
-      }
+      values = getFormattedTimeClause(clause, timezone);
       break;
-
     default:
       throw new Error(`unknown kind ${dimKind}`);
   }
 
-  return {title: getClauseLabel(), values};
+  return { title: getClauseLabel(), values };
+}
+
+export function getFormattedTimeClause(clause: FilterClause, timezone: Timezone): string {
+  let values;
+
+  if (clause.relative) {
+    values = getRelativeTimeLabel(clause);
+  } else {
+    values = getFixedTimeRangeLabel(clause, timezone);
+  }
+
+  return values;
 }
 
+function getRelativeTimeLabel(clause: FilterClause): string {
+  const { selection } = clause;
+
+  if (selection instanceof TimeRangeExpression) {
+    return getRelativeTimeRangeLabel(selection);
+  } else if (selection instanceof TimeBucketExpression) {
+    return getRelativeTimeBucketLabel(selection);
+  } else {
+    throw Error(`unsupported relative time filter clause: ${clause.selection}`);
+  }
+}
+
+function getRelativeTimeRangeLabel(expression: TimeRangeExpression): string {
+  const op = expression.getHeadOperand();
+  let durationDescription = expression.getQualifiedDurationDescription();
+  if (expression.duration.toString() === 'P3M') {
+    durationDescription = "quarter";
+  }
+
+  if (op.equals($(FilterClause.MAX_TIME_REF_NAME)) && expression.step < 0) {
```

Comment:
what does expression.step ... mean? encapsulate expression in variable or function